File: journal-service/src/main/java/com/example/journal/service/SubjectService.java
package com.example.journal.service;

import com.example.journal.model.GroupHasSubject;
import com.example.journal.model.Subject;
import com.example.journal.repository.GroupHasSubjectRepository;
import com.example.journal.repository.SubjectRepository;
import java.util.List;
import java.util.Optional;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class SubjectService {
    private final SubjectRepository subjectRepository;
    private final GroupHasSubjectRepository groupHasSubjectRepository;

    public Optional<Subject> getByName(final String name) {
        return subjectRepository.getByName(name);
    }

    public Subject save(Subject subject) {
        return subjectRepository.save(subject);
    }

    public Optional<Subject> getById(final Long subjectId) {
        return subjectRepository.getById(subjectId);
    }

    public GroupHasSubject addSubjectToTheGroup(final GroupHasSubject groupHasSubject) {
        return groupHasSubjectRepository.save(
                groupHasSubject.getGroupId(),
                groupHasSubject.getTeacherId(),
                groupHasSubject.getSubjectId()
        );
    }

    public List<Subject> getGroupSubjects(final Long groupId) {
        return subjectRepository.getGroupListSubject(groupId);
    }

    public List<GroupHasSubject> getTeacherSubjects(final Long teacherId) {
        return groupHasSubjectRepository.getByTeacherId(teacherId);
    }

    public List<Subject> getSubjectsByTeacherIdAndGroupId(Long teacherId, Long groupId) {
        return subjectRepository.getSubjectsByTeacherIdAndGroupId(teacherId, groupId);
    }

    public List<Subject> getSubjectList() {
        return subjectRepository.getSubjectList();
    }
}

File: ui/src/service/SubjectService.ts
import axios from "axios";
import {GroupHasSubject, Subject} from "../model/SubjectState";

class SubjectService {
    getGroupSubjects = (id: number | undefined): Promise<Subject[]> => {
        return axios.get<Subject[]>(
            `http://localhost:8080/api/v1/group/${id}/subjects`,
            {withCredentials: true}
        ).then(subjects => {
            if (subjects.data.length != null) {
                return subjects.data;
            } else {
                return [];
            }
        });

    }

    getTeacherSubjects = (id: number): Promise<GroupHasSubject[]> => {
        return axios.get<GroupHasSubject[]>(
            `http://localhost:8080/api/v1/teacher/${id}/subjects`,
            {withCredentials: true}
        ).then(request => request.data);
    }

    getSubjectsByTeacherIdAndGroupId = (teacherId: number, groupId: number): Promise<Subject[]> => {
        return axios.get<Subject[]>(
            `http://localhost:8080/api/v1/teacher/${teacherId}/group/${groupId}/subjects`,
            {withCredentials: true}
        ).then(request => request.data);
    }
}

export default new SubjectService();
